modules/gdscript/gdscript_formatter.cpp

```diff
@@ -0,0 +1,231 @@
+			return false;
+		}
+
+		// )(
+		if (cur == GDScriptTokenizer::TK_PARENTHESIS_CLOSE) {
+			return false;
+		}
+	}
+
+	if (right == GDScriptTokenizer::TK_PARENTHESIS_CLOSE) {
+		return false;
+	}
+
+	// Brackets
+
+	if (cur == GDScriptTokenizer::TK_BRACKET_OPEN) {
+		return false;
+	}
+
+	if (right == GDScriptTokenizer::TK_BRACKET_OPEN) {
+		// Check whether we are indexing into an expression, or we have the start of a list literal.
+		// We should not insert a space in the former case, but we should in the latter.
+		return !is_end_of_expression(cur);
+	}
+
+	if (right == GDScriptTokenizer::TK_BRACKET_CLOSE) {
+		return false;
+	}
+
+	// Curly brackets
+
+	if (cur == GDScriptTokenizer::TK_CURLY_BRACKET_OPEN) {
+		// We always want a space after a curly bracket, unless it's immediately followed by a
+		// closing culy bracket.
+		return right != GDScriptTokenizer::TK_CURLY_BRACKET_CLOSE;
+	}
+
+	// Unary operators
+
+	if (cur == GDScriptTokenizer::TK_OP_BIT_INVERT) {
+		return false;
+	}
+
+	if (cur == GDScriptTokenizer::TK_DOLLAR) {
+		return false;
+	}
+
+	if (cur == GDScriptTokenizer::TK_OP_SUB) {
+		// Check if we have a unary minus sign by looking at the previous token.
+		// If the previous token is the end of an expression, then we have a binary minus.
+		// Otherwise, we have a unary minus.
+		return is_end_of_expression(left);
+	}
+
+	// Other separators
+
+	if (right == GDScriptTokenizer::TK_COMMA) {
+		return false;
+	}
+
+	if (cur == GDScriptTokenizer::TK_PERIOD || right == GDScriptTokenizer::TK_PERIOD) {
+		return false;
+	}
+
+	if (right == GDScriptTokenizer::TK_COLON) {
+		return false;
+	}
+
+	return true;
+}
+} // namespace
+
+GDScriptFormatter::GDScriptFormatter(const String &source) :
+		_source(&source[0]),
+		_tokenizer(memnew(GDScriptTokenizerText)) {
+	_tokenizer->set_code(source);
+}
+
+GDScriptFormatter::~GDScriptFormatter() {
+	memdelete(_tokenizer);
+}
+
+bool GDScriptFormatter::format() {
+	while (_tokenizer->get_token() != GDScriptTokenizer::TK_EMPTY) {
+		// Comments and '\\' (and any whitespace afterwards) are skipped by the tokenizer, so in order to retain them,
+		// check the 'between_text' for non-whitespace, then if there is any, copy all of it.
+		int prev_end = 0;
+		if (_tokenizer->get_token(-1) != GDScriptTokenizer::TK_EMPTY) {
+			prev_end = _tokenizer->get_token_code_pos(-1) + _tokenizer->get_token_code_len(-1);
+		}
+
+		int start = _tokenizer->get_token_code_pos();
+		String between_text;
+		for (int i = prev_end; i < start; i++) {
+			between_text += _source[i];
+		}
+		if (between_text.strip_edges().length() > 0) {
+			_formatted += between_text;
+		}
+
+		switch (_tokenizer->get_token()) {
+			case GDScriptTokenizer::TK_EOF:
+				return true;
+			default: {
+				// Special case for unary '!' operator which requires no space after it
+				if (_tokenizer->get_token() == GDScriptTokenizer::TK_EOF && _source[_tokenizer->get_token_code_pos()] == '!') {
```

Comment:
Dead code, `_tokenizer->get_token() == GDScriptTokenizer::TK_EOF` cannot be true.